src/parquet/bloom_filter_algorithm.cc

```diff
@@ -0,0 +1,102 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "parquet/bloom_filter_algorithm.h"
+#include "arrow/status.h"
+#include "arrow/util/bit-util.h"
+#include "parquet/exception.h"
+#include "parquet/types.h"
+
+namespace parquet {
+constexpr uint32_t BlockBasedAlgorithm::SALT[BITS_SET_PER_BLOCK];
+
+BlockBasedAlgorithm::BlockBasedAlgorithm(uint32_t num_bytes)
+    : pool_(::arrow::default_memory_pool()) {
+  if ((num_bytes & (num_bytes - 1)) != 0) {
+    throw ParquetException("num_bytes is not a power of 2");
+  }
+
+  num_bytes_ = num_bytes;
+
+  ::arrow::Status status =
```

Comment:
Instead of calling raw `Allocate` use https://github.com/apache/arrow/blob/c6e33d89d10d9aa324cbd5b6776cadbb367970a7/cpp/src/arrow/buffer.h#L232 to get a Buffer. This will then be automatically be deallocated on destruction of the reference.

Reply:
Thanks, I will use it. Just one thing needs to be consider, converting from uint8_t* to uint32_t* with reinterpret_cast may lead to undefined behavior.

Hi @jbapple-cloudera , looks like reinterpret_cast from uint8_t* to other bigger size type pointer is widely used in this project and also arrow, such as TypedBufferBuilder. Is the undefined behavior not present when the pointer's address and the buffer size is a multiply of requirement?

Reply:
https://github.com/apache/parquet-cpp/pull/425 removed `-fno-strict-aliasing` from the build flags. This makes many common calls to `reinterpret_cast` invalid. If you see lots of other use, I'd suggest starting a thread on dev@ on the subject.